### PR TITLE
turbo type-check command and GHA

### PR DESCRIPTION
**Pull request by anthonyshew**
## What kind of change does this PR introduce?

DX feature!

## What is the current behavior?

Currently, there's a `tsc` check at the root of the repo for type checking this monorepo.

This works! But, with `turbo` in the mix...

## What is the new behavior?

Type checking the monorepo `turbo` style! This incremental adoption shouldn't have too much impact on the repo by itself - but does open us up to surface and fix type errors more easily!

- Parallelized `tsc` per workspace
- Caching once we start having successful runs

I've also add a GitHub Action to this PR if the core team would like to use it for checks. Completely optional, of course!

## Additional context

This PR only adds the type checking commands but does not resolve the errors that it surfaces. With this change, we'll find many opportunities for better codebase health. Incremental adoption with followup PRs to clear errors is best in this case! 😄

### CODE REVIEW

**Comment by joshenlim** [comment]
hey there @anthonyshew! thanks so much for this :) just to check real quick though - the TS type check seems to be failing here with the following message

`Invalid tag name "tsc --noEmit" of package "type-check@tsc --noEmit": Tags may not have any characters that encodeURIComponent encodes`

**Comment by joshenlim** [comment]
hey there @anthonyshew! thanks so much for this - excited for these DX improvements! 😄 
however i'm afraid that we might have to resolve those TS check errors and getting the checks to pass first as the errors are also causing the [builds](https://vercel.com/supabase/supabase-studio-staging/2VqEyCSMggihW7hyGgFbzUi7Qx2R) to fail

File: packages/common/package.json
Comment (ruggi99) [inline packages/common/package.json:10]: This should be in the `scripts` field

Comment (anthonyshew) [reply packages/common/package.json:10]: Well, that's embarrassing. 😄

**Comment by anthonyshew** [comment]
@joshenlim That's true, I didn't really know what path to take as far as clearing or ignoring errors that would work best for y'all on core team. `turbo` is plenty incrementally adoptable so you've got options:

1. Clear the errors (I wouldn't recommend this, it could be high impact and difficult to get merged in)
2. Change task names in your workspaces to not be a part of `turbo run type-check` by renaming them to `IGNORED_type-check` (or something else that isn't `type-check`)
3. Remove the offending `type-check` command completely and look to add more type checking in the future

There are other strategies to go with but those tend to be the most simple!

File: packages/common/package.json
Comment (ruggi99) [inline packages/common/package.json:7]:
```suggestion
  "scripts": {
```

**Comment by joshenlim** [comment]
hey there @anthonyshew! apologies for the delay on this - just managed to take a look proper now 🙏 
think what you mentioned is a great idea and we'll probably just resolve the typescript errors incrementally as we go along

i'm keen to get this PR in as a best practice of using turbo, but before that, could we resolve the GH actions that are failing? We'll need to ensure that they're all passing minimally before we merge the PR:
- Check code with Prettier
  - Would you mind getting the changes from master into this branch? Previously there was an issue with review dog that was causing the action to fail so we've removed it
- Check code with Typescript
  - I'm seeing a non typescript check related error [here](https://github.com/supabase/supabase/actions/runs/4033560178/jobs/7266560445?pr=11658) - wondering if this has anything to do with the way its currently set up? or is this coming from a specific app?
image
- Vercel builds
  - I'm checking `studio-staging` specifically but the deployment is failing with an odd error, perhaps i'll check on this again after we get the changes from master into this branch!
image

**Comment by anthonyshew** [comment]
Those errors look like the package-lock got mangled somehow. I've brought main in, hopefully that gives it a fix. 👍

**Comment by ivasilov** [comment]
hey @anthonyshew, thank you for your work so far. Can you update this PR so that we can merge it? I can also finish it if you can't find the time.

File: .github/workflows/quality.yml
Comment (kevcodez) [inline .github/workflows/quality.yml:1]: Nit: Rename the file to `typecheck.yml`

**Review by kevcodez — APPROVED** [review]

**Review by ivasilov — APPROVED** [review]

**Review by alaister — APPROVED** [review]

**Pull request merged**
